plutus-core/plutus-ir/src/PlutusIR/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,148 @@
+mapBranches ::
+  forall tyname name uni fun a
+  . (Term tyname name uni fun a -> Maybe (Term tyname name uni fun a))
+  -> AppContext tyname name uni fun a
+  -> [Arity]
+  -> Maybe (AppContext tyname name uni fun a)
+mapBranches f = go
+    where
+    go :: AppContext tyname name uni fun a -> [Arity] -> Maybe (AppContext tyname name uni fun a)
+    go AppContextEnd [] = Just AppContextEnd
+    go (TermAppContext branch ann ctx) (arity:arities) =
+      TermAppContext <$> underBindersMaybe arity f branch <*> pure ann <*> go ctx arities
+    go _ _ = Nothing
+
+    -- I tried to come up with a way to do this in terms of the underBinders traversal
+    -- but I couldn't :(
+    underBindersMaybe
+      :: Arity
+      -> (Term tyname name uni fun a -> Maybe (Term tyname name uni fun a))
+      -> Term tyname name uni fun a
+      -> Maybe (Term tyname name uni fun a)
+    underBindersMaybe [] g t                                = g t
+    underBindersMaybe (TermParam:arity) g (LamAbs a n ty t) = LamAbs a n ty <$> underBinders arity g t
+    underBindersMaybe (TypeParam:arity) g (TyAbs a ty k t)  = TyAbs a ty k <$> underBinders arity g t
+    underBindersMaybe _ _ _                                 = Nothing
+
+-- | Split a match, either a normal datatype match or a builtin 'match'.
+splitMatch :: (Ord fun, PLC.HasUnique name PLC.TermUnique, PLC.HasUnique tyname PLC.TypeUnique) =>
+  BuiltinsInfo uni fun
+  -> VarsInfo tyname name
+  -> Term tyname name uni fun a
+  -> Maybe (SplitMatchContext tyname name uni fun a, [Arity], SplitMatchContext tyname name uni fun a -> Term tyname name uni fun a)
+splitMatch binfo vinfo t = do
+  let (hd, args) = splitApplication t
+  (sm, arities, reconstruct) <- case hd of
+    (Var _ matcherName)     -> splitNormalDatatypeMatch vinfo matcherName args
+    (Builtin _ matcherName) -> splitBuiltinDatatypeMatch binfo matcherName args
+    _                       -> Nothing
+  -- wrangle the reconstruction function so it also puts the head back in, for
+  -- convenience of use elsewhere
+  pure (sm, arities, \sm' -> fillAppContext hd (reconstruct sm'))
+
+{- Note [Case-of-case and conapps]
+Case-of-case is *guaranteed* to be good if the bodies of the branches are all constructor applications.
+Consider the following program we might typically get from Plutus Tx:
+
+Bool_match {integer} (ifThenElse {Bool} (lessThanInteger 1 2) True False) 3 4
+---> case-of-case
+ifThenElse {integer} (lessThanInteger 1 2) (Bool_match {integer} True 3) (Bool_match {integer} False 4)
+---> case-of-known-constructor
+ifThenElse {integer} (lessThanInteger 1 2) 3 4
+
+That is, this guarantees that case-of-known constructor will fire and get rid of one of the
+matches entirely, which is great.
+
+If we *don't* have this property, then case-of-case can duplicate code, making the program bigger.
+So the conservative option is to only do case-of-case when this is true.
```

Comment:
Aha, here they are. Is case-of-case interleaved with case-of-known-constructor? If not, can there still be an exponential blowup in code size before case-of-known-constructor cleans everything up? Worth discussing in this Note I think.

Reply:
Also note that I'm currently turning the non-conservative version on by default since it seems to help! I feel pretty nervous about that...

Good point about the interleaving: indeed, they run one after the other and only once in each simplifier iteration, so we should be safe, but worth pointing out.

Reply:
Good point, that seems possible. Let me try and come up with a case where that could happen.

Reply:
Yep, I can make it happen. I think if we're in conservative mode then COKC is still guaranteed to clear it up, so the worst that can happen is a big intermediate tree.